lib/logger/src/lib.rs

```diff
@@ -0,0 +1,102 @@
+#[macro_use]
+extern crate log;
+
+use std::cell::RefCell;
+use std::fs::File;
+use std::io::{self, Write};
+use std::sync::atomic;
+use std::thread;
+
+use log::{Level, LevelFilter, Metadata, Record};
+
+static STATE: atomic::AtomicIsize = atomic::ATOMIC_ISIZE_INIT;
+
+thread_local! {
+    static WRITER: RefCell<Option<io::BufWriter<File>>> = RefCell::new(None);
+}
+
+pub fn initialize(filename_prefix: &str, enable: bool) {
+    // Ensure the thread local state is always properly initialized.
+    WRITER.with(|rc| {
+        if rc.borrow().is_none() {
+            rc.replace(Some(open_file(filename_prefix)));
+        }
+    });
+
+    let new_state = match enable {
+        true => 1,
+        false => -1,
+    };
+
+    let previous_state = STATE.load(atomic::Ordering::Relaxed);
+    if previous_state == new_state {
+        return;
+    }
+
+    // Once and only once in the entire program, set up the logger.
+    if previous_state == 0 {
+        log::set_boxed_logger(Box::new(ThreadPerFileLogger))
+            .map(|()| log::set_max_level(LevelFilter::Info))
+            .expect("setting up logging panicked");
+    }
+
+    STATE.store(new_state, atomic::Ordering::Relaxed);
+    info!(
+        "Set up logging to {:?}; filename prefix is {}",
+        enable, filename_prefix
+    );
+}
+
+#[inline]
+fn enabled() -> bool {
+    STATE.load(atomic::Ordering::Relaxed) > 0
+}
```

Comment:
In the old code, this had a check for `cfg!(debug_assertions)` and just returned false if they weren't enabled. Is there a reason for not doing that here?

Reply:
Also, if we added that here, do you know if that would be enable to allow debugging code to be optimized out in release builds?

Reply:
First question: I moved this in the code that initializes the lib in codegen, because it seemed like a Cranelift choice, not something any user of this library would want in general.

Second question: it seems not, but there's a macro for doing so, provided by `log`: [log_enabled](https://docs.rs/log/0.4.3/log/macro.log_enabled.html)